KHR_materials_dispersion (#2340)
KhronosGroup/glTF: extensions/2.0/Khronos/KHR_materials_dispersion/README.md

## Status

Draft

## Dependencies

- Written against the glTF 2.0 spec.
- The `KHR_materials_volume` extension as it builds upon its volumetric effect.

## Exclusions

- This extension must not be used on a material that also uses `KHR_materials_pbrSpecularGlossiness`.
- This extension must not be used on a material that also uses `KHR_materials_unlit`.

## Overview

This extension adds one parameter to the metallic-roughness material: `dispersion`.

`dispersion` enables configuring the strength of the angular separation of colors (chromatic aberration) transmitting through a relatively clear volume.  It is an enhancement to the default ```KHR_materials_volume``` transmission model which assumes no dispersion.

Optical dispersion is represented in terms of the Abbe number parameterization \( $V$ \).  The dispersion effect is a result of the wavelength-dependent index of refraction of a material.  Dispersion is a widely adopted parameter in modern PBR models.  It is present in both OpenPBR (as *transmission_dispersion_abbe_number*) and the Dassault Enterprise PBR Shading Model (as $V_d$).

> The Abbe number \( $V$ \) is computed from the index of refraction at three wavelengths of visible light: 486.1 nm (short wavelength blue, $n_F$), 587.6 nm (central wavelength, $n_d$), and 656.3 nm (long wavelength red, $n_C$).  The Abbe number makes the simplifying assumption that the index of refraction variance is linear:

Reviewer: The TSG agreed on the $V_d$ change. Done.

$$V = \frac{n_d - 1}{n_F - n_C}$$

> To calculate the index of refraction at a specific wavelength \( $\lambda$ \), given an Abbe number \( $V$ \) and the central index of refraction as specified by the ```KHR_materials_ior``` extension (assumed to be at the central wavelength, \( $N_d$ \)):

Reviewer: Thanks @proog128.  Great to see you here!

$$
B = \frac{n_d - 1}{V \left( {\lambda_F^{-2}} - {\lambda_C^{-2}} \right)}
$$

And then:

$$
A = n_d - \frac{B}{\lambda_d^2}
$$

And finally:

$$

> n(\lambda) = A + \frac{B}{\lambda^2}

Reviewer 2: To be precise, the first two terms of the general Cauchy equation. We should probably add this info as well?
$$
\eta(\lambda) = a + \frac{b}{\lambda^2} + \frac{c}{\lambda^4} + ...,
$$

> n(\lambda) = A + \frac{B}{\lambda^2}

Reviewer: I'm unsure how to actually add this to the draft.  Can someone write up the exact change(s) wanted here?

$$

![Dispersion on a Gem](./figures/Dispersion.jpg)

> In this extension, we store a transformed dispersion instead of the Abbe number directly.  Specifically we store $20/V$ so that a value of 1.0 is equivalent to $V=20$, which is about the lowest Abbe number for normal materials. Values over 1.0 are still valid for artists that want to exaggerate the effect. Decreasing values lower the amount of dispersion down to 0.0.  This is the same transform used by both Adobe Standard Material and OpenPBR.

Reviewer: The group talked about this, but decided that this mapping is in common use in other systems (Adobe ASM, OpenPBR has both this and raw Abbe available, etc).  For people looking up physical values in a table, this is a simple calculation to apply, and doesn't make it any less "physical."  For example one may convert Fahrenheit to Celsius without loosening the relationship with a physical temperature.  And for artists wishing to take control with creative license, having a simple linear control is clearly the better choice.  So, I think we will leave $20/V_d$ in place.

Here is a table of some material dispersion Abbe numbers, including the outlier of Rutile which is a very high dispersion material:

| Material | Abbe Number (V) | Dispersion (20/V) |
| -------- | ----------- | -------------- |
| Rutile | 9.8 | 2.04 |
| Polycarbonate | 32 | 0.625 |
| Diamond | 55 | 0.36 |
| Water | 55 | 0.36 |
| Crown Glass | 59 | 0.33 |

## Implementation

*This section is non-normative.*

One real-time method for rendering dispersion effects is to trace volume transmission separately for each of color channel accounting for the per channel IOR as determined by the Abbe number.  The resulting composite image will show color separation between the channels as a result.

## Schema

- [material.KHR_materials_dispersion.schema.json](schema/material.KHR_materials_dispersion.schema.json)

## Reference

### Theory, Documentation and Implementations

[Abbe Number - Wikipedia](https://en.wikipedia.org/wiki/Abbe_number)

> [Abbe Number - Wolfram Formula Repository](https://resources.wolframcloud.com/FormulaRepository/resources/Abbe-Number#:~:text=The%20Abbe%20number%2C%20also%20known,of%20V%20indicating%20low%20dispersion.)

Reviewer 3: Consider dropping the fragment part (starts with `#`) of the link.

Reviewer: Fixed.

## Extending Materials

The dispersion, defined in terms of Abbe number, is defined by adding the `KHR_materials_dispersion` extension to any glTF material.

```json
{
    "materials": [
        {
            "extensions": {
                "KHR_materials_dispersion": {
                    "dispersion": 0.1
                }
            }
        }
    ]
}
```

Factor and texture are combined by multiplication to describe a single value.

| |Type|Description|Required|
|-|----|-----------|--------|
| **dispersion** | `number` | The strength of the dispersion effect, specified as 20/Abbe number. | No, default: `0`|

The default value of 0 has a special meaning in that no dispersion should be used.  This is the default value for backwards compatibility.  Any value zero or larger is considered to be a valid dispersion value, although the range between [0 , 1] is the range of realistic values.

Here is a table of some material dispersion Abbe numbers, including the outlier of Rutile which is a very high dispersion material:

> | Material | Abbe Number (V) |

Author: addressed

## Status

Draft

## Dependencies

- Written against the glTF 2.0 spec.
- The `KHR_materials_volume` extension as this builds upon its volumetric effect.

## Exclusions

- This extension must not be used on a material that also uses `KHR_materials_pbrSpecularGlossiness`.
- This extension must not be used on a material that also uses `KHR_materials_unlit`.

## Overview

This extension adds one parameter to the metallic-roughness material: `dispersion`.

`dispersion` enables configuring the strength of the angular separation of colors (chromatic aberration) transmitting through a relatively clear volume.  It is an enhancement to the default ```KHR_materials_volume``` transmission model which assumes no dispersion.

Optical dispersion is represented in terms of the Abbe number parameterization \( $V$ \).  The dispersion effect is a result of the wavelength-dependent index of refraction of a material.  Dispersion is a widely adopted parameter in modern PBR models.  It is present in both OpenPBR (as *transmission_dispersion_abbe_number*) and the Dassault Enterprise PBR Shading Model (as $V_d$).

> The Abbe number \( $V$ \) is computed from the index of refraction at three wavelengths of visible light: 486.1 nm (short wavelength blue, $n_F$), 587.6 nm (central wavelength, $n_d$), and 656.3 nm (long wavelength red, $n_C$).  The Abbe number makes the simplifying assumption that the index of refraction variance is linear:

Reviewer 2: Suggest small reformulation to clarify that this is the Abbe definition for visible light (others exist), and to add info on the origin of the three wavelength constants. 

```suggestion
The Abbe number \( $V$ \) for visible light is computed from the index of refraction at three wavelengths of Fraunhofer's spectral lines: 486.1 nm (short wavelength blue, $n_F$), 587.6 nm (central wavelength, $n_d$), and 656.3 nm (long wavelength red, $n_C$). The Abbe number makes the simplifying assumption that the index of refraction variance is linear:
```

## Overview

This extension adds one parameter to the metallic-roughness material: `dispersion`.

`dispersion` enables configuring the strength of the angular separation of colors (chromatic aberration) transmitting through a relatively clear volume.  It is an enhancement to the default ```KHR_materials_volume``` transmission model which assumes no dispersion.

Optical dispersion is represented in terms of the Abbe number parameterization \( $V_d$ \).  The dispersion effect is a result of the wavelength-dependent index of refraction of a material.  Dispersion is a widely adopted parameter in modern PBR models.  It is present in both OpenPBR (as *transmission_dispersion_abbe_number*) and the Dassault Enterprise PBR Shading Model (as $V_d$).

The Abbe number \( $V_d$ \) for visible light is computed from the index of refraction at three wavelengths of Fraunhofer's spectral lines: 486.1 nm (short wavelength blue, $n_F$), 587.6 nm (central wavelength, $n_d$), and 656.3 nm (long wavelength red, $n_C$). The Abbe number makes the simplifying assumption that the index of refraction variance is linear:

$$V_d = \frac{n_d - 1}{n_F - n_C}$$

> To calculate the index of refraction at a specific wavelength \( $\lambda$ \), given an Abbe number \( $V_d$ \) and the central index of refraction as specified by the ```KHR_materials_ior``` extension (assumed to be at the central wavelength, \( $n_d$ \)):
> 
> $$
> B = \frac{n_d - 1}{V_d \left( {\lambda_F^{-2}} - {\lambda_C^{-2}} \right)}
> $$
> 
> And then:
> 
> $$
> A = n_d - \frac{B}{\lambda_d^2}
> $$
> 
> And finally:
> 
> $$
> n(\lambda) = A + \frac{B}{\lambda^2}

Reviewer 2: > I'm unsure how to actually add this to the draft. Can someone write up the exact change(s) wanted here?

```suggestion
To calculate the index of refraction at a specific wavelength \( $\lambda$ \), we use the first two terms of Cauchy's equation:

$$
n(\lambda) = A + \frac{B}{\lambda^2}
$$

Given an Abbe number \( $V_d$ \) and the central index of refraction as specified by the ```KHR_materials_ior``` extension (assumed to be at the central wavelength, \( $n_d$ \)), the coefficients are defined as:

$$
B = \frac{n_d - 1}{V_d \left( {\lambda_F^{-2}} - {\lambda_C^{-2}} \right)}
$$

And:

$$
A = n_d - \frac{B}{\lambda_d^2}
$$

By substituting the symbolic wavelength constants $n_d$, $n_F$ and $n_C$ with the values provided above we get:

$$
n(\lambda) = \max(n_d + \frac{n_d-1}{V_d} \left( \frac{523655}{\lambda^2} - 1.5168 \right), 1)

```

Reviewer: Thanks!

The Abbe number \( $V_d$ \) for visible light is computed from the index of refraction at three wavelengths of Fraunhofer's spectral lines: 486.1 nm (short wavelength blue, $n_F$), 587.6 nm (central wavelength, $n_d$), and 656.3 nm (long wavelength red, $n_C$). The Abbe number makes the simplifying assumption that the index of refraction variance is linear:

$$V_d = \frac{n_d - 1}{n_F - n_C}$$

To calculate the index of refraction at a specific wavelength \( $\lambda$ \), we use the first two terms of Cauchy's equation:

$$
n(\lambda) = A + \frac{B}{\lambda^2}
$$

Given an Abbe number \( $V_d$ \) and the central index of refraction as specified by the ```KHR_materials_ior``` extension (assumed to be at the central wavelength, \( $n_d$ \)), the coefficients are defined as:

$$
B = \frac{n_d - 1}{V_d \left( {\lambda_F^{-2}} - {\lambda_C^{-2}} \right)}
$$

And:

$$
A = n_d - \frac{B}{\lambda_d^2}
$$

By substituting the symbolic wavelength constants $n_d$, $n_F$ and $n_C$ with the values provided above we get:

$$
n(\lambda) = \max(n_d + \frac{n_d-1}{V_d} \left( \frac{523655}{\lambda^2} - 1.5168 \right), 1)
$$

> In this extension, we store a transformed dispersion instead of the Abbe number directly.  Specifically we store $20/V_d$ so that a value of 1.0 is equivalent to $V_d=20$, which is about the lowest Abbe number for normal materials. Values over 1.0 are still valid for artists that want to exaggerate the effect. Decreasing values lower the amount of dispersion down to 0.0.  This is the same transform used by both Adobe Standard Material and OpenPBR.

Reviewer 3: Please put the last sentence into a NOTE block as it's not normative.

```
> [!NOTE]  
> This is the same transform ...
```

Reviewer: Added.  This didn't display in VSCode's markdown preview, but works fine on GitHub.

The Abbe number \( $V_d$ \) for visible light is computed from the index of refraction at three wavelengths of Fraunhofer's spectral lines: 486.1 nm (short wavelength blue, $n_F$), 587.6 nm (central wavelength, $n_d$), and 656.3 nm (long wavelength red, $n_C$). The Abbe number makes the simplifying assumption that the index of refraction variance is linear:

$$V_d = \frac{n_d - 1}{n_F - n_C}$$

To calculate the index of refraction at a specific wavelength \( $\lambda$ \), we use the first two terms of Cauchy's equation:

$$
n(\lambda) = A + \frac{B}{\lambda^2}
$$

Given an Abbe number \( $V_d$ \) and the central index of refraction as specified by the ```KHR_materials_ior``` extension (assumed to be at the central wavelength, \( $n_d$ \)), the coefficients are defined as:

$$
B = \frac{n_d - 1}{V_d \left( {\lambda_F^{-2}} - {\lambda_C^{-2}} \right)}
$$

And:

$$
A = n_d - \frac{B}{\lambda_d^2}
$$

By substituting the symbolic wavelength constants $n_d$, $n_F$ and $n_C$ with the values provided above we get:

$$
n(\lambda) = \max(n_d + \frac{n_d-1}{V_d} \left( \frac{523655}{\lambda^2} - 1.5168 \right), 1)

> $$

Reviewer 3: This block does not render on GitHub. Maybe try removing the empty line?

Reviewer: Fixed, thanks!

## Status

Release Candidate

## Dependencies

- Written against the glTF 2.0 spec.
- The `KHR_materials_volume` extension as this builds upon its volumetric effect.

## Exclusions

- This extension must not be used on a material that also uses `KHR_materials_pbrSpecularGlossiness`.
- This extension must not be used on a material that also uses `KHR_materials_unlit`.

## Overview

This extension adds one parameter to the metallic-roughness material: `dispersion`.

`dispersion` enables configuring the strength of the angular separation of colors (chromatic aberration) transmitting through a relatively clear volume.  It is an enhancement to the default ```KHR_materials_volume``` transmission model which assumes no dispersion.

Optical dispersion is represented in terms of the Abbe number parameterization \( $V_d$ \).  The dispersion effect is a result of the wavelength-dependent index of refraction of a material.  Dispersion is a widely adopted parameter in modern PBR models.  It is present in both OpenPBR (as *transmission_dispersion_abbe_number*) and the Dassault Enterprise PBR Shading Model (as $V_d$).

> The Abbe number \( $V_d$ \) for visible light is computed from the index of refraction at three wavelengths of Fraunhofer's spectral lines: 486.1 nm (short wavelength blue, $n_F$), 587.6 nm (central wavelength, $n_d$), and 656.3 nm (long wavelength red, $n_C$). The Abbe number makes the simplifying assumption that the index of refraction variance is linear:
> 
> $$V_d = \frac{n_d - 1}{n_F - n_C}$$

Reviewer 2: @lexaknyazev  
- Change wording wrt. Fraunhofer spectral lines
- Define (more precise) lambda symbols 

```suggestion
With the simplifying assumption that the index of refraction variance is linear,
the Abbe number \( $V_d$ \) for visible light is computed as:
$$V_d = \frac{n_d - 1}{n_F - n_C}$$

where $n_F$, $n_d$ and $n_C$ are the indices of refraction at the wavelengths of the Fraunhofer F, d, and C spectral lines: 
$$\lambda_F = 486.13 nm$$
$$\lambda_d = 587.56 nm$$
$$\lambda_C = 656.27 nm$$

```

To calculate the index of refraction at a specific wavelength \( $\lambda$ \), we use the first two terms of Cauchy's equation:

$$
n(\lambda) = A + \frac{B}{\lambda^2}
$$

Given an Abbe number \( $V_d$ \) and the central index of refraction as specified by the ```KHR_materials_ior``` extension (assumed to be at the central wavelength, \( $n_d$ \)), the coefficients are defined as:

$$
B = \frac{n_d - 1}{V_d \left( {\lambda_F^{-2}} - {\lambda_C^{-2}} \right)}
$$

And:

$$
A = n_d - \frac{B}{\lambda_d^2}
$$

> By substituting the symbolic wavelength constants $n_d$, $n_F$ and $n_C$ with the values provided above we get:

Reviewer 3: This should likely say `\lambda_d`, `\lambda_F`, and `\lambda_C`.

$$

> n(\lambda) = \max(n_d + \frac{n_d-1}{V_d} \left( \frac{523655}{\lambda^2} - 1.5168 \right), 1)

Reviewer 2: @lexaknyazev With the changed precision of the lambdas the numbers should be ok now.

Optical dispersion is represented in terms of the Abbe number parameterization \( $V_d$ \).  The dispersion effect is a result of the wavelength-dependent index of refraction of a material.  Dispersion is a widely adopted parameter in modern PBR models.  It is present in both OpenPBR (as *transmission_dispersion_abbe_number*) and the Dassault Enterprise PBR Shading Model (as $V_d$).

With the simplifying assumption that the index of refraction variance is linear,
the Abbe number \( $V_d$ \) for visible light is computed as:
$$V_d = \frac{n_d - 1}{n_F - n_C}$$

where $n_F$, $n_d$ and $n_C$ are the indices of refraction at the wavelengths of the Fraunhofer F, d, and C spectral lines: 
$$\lambda_F = 486.13 nm$$
$$\lambda_d = 587.56 nm$$
$$\lambda_C = 656.27 nm$$

To calculate the index of refraction at a specific wavelength \( $\lambda$ \), we use the first two terms of Cauchy's equation:

$$
n(\lambda) = A + \frac{B}{\lambda^2}
$$

> Given an Abbe number \( $V_d$ \) and the central index of refraction as specified by the ```KHR_materials_ior``` extension (assumed to be at the central wavelength, \( $n_d$ \)), the coefficients are defined as:

Reviewer 4: `\( $n_d$ \))` -> `\( $n_d$ \)`

Reviewer 3: Why? The closing parenthesis is needed to balance `(assumed`.

Reviewer: Is there any easy way to avoid nested parenthesis here?  Or should we just leave it?

Reviewer 3: We could just remove all parentheses around in-text variables. It's an editorial choice.

Reviewer 4: Right, sorry. Looked awkward at first, but it's correct.

> # KHR\_materials\_disperson

Reviewer 3: nit: typo

Author: fixed
